_posts/2023-05-17-Scaling_part_2.md

```diff
@@ -0,0 +1,642 @@
+However, that is not how the API server works.
+It expects to be able to authenticate using client certificates.
+Luckily, etcd supports this so we just have to generate the certificates and sign them so that etcd trusts them.
+The key thing is to set the common name in the certificate to the name of the user we want to authenticate as.
+
+Since `kubeadm` always sets the same common name, we will here use `openssl` to generate the client certificates so that we get control over it.
+
+```bash
+# Generate etcd client certificate
+openssl req -newkey rsa:2048 -nodes -subj "/CN=test" \
+ -keyout "/tmp/test/pki/apiserver-etcd-client.key" -out "/tmp/test/pki/apiserver-etcd-client.csr"
+openssl x509 -req -in "/tmp/test/pki/apiserver-etcd-client.csr" \
+  -CA /tmp/test/pki/etcd/ca.crt -CAkey /tmp/test/pki/etcd/ca.key -CAcreateserial \
+  -out "/tmp/test/pki/apiserver-etcd-client.crt" -days 365
+```
+
+### Deploying the API server
+
+In order to deploy the API server, we will first need to generate some more certificates.
+The client certificates for connecting to etcd are already ready, but it also needs certificates to secure the exposed API itself, and a few other things.
+Then we will also need to create secrets from all of these certificates:
+
+```bash
+kubeadm init phase certs ca --config kubeadm-config.yaml
+kubeadm init phase certs apiserver --config kubeadm-config.yaml
+kubeadm init phase certs sa --cert-dir /tmp/test/pki
+
+kubectl create ns workload-api
+kubectl -n workload-api create secret tls test-ca --cert /tmp/test/pki/ca.crt --key /tmp/test/pki/ca.key
+kubectl -n workload-api create secret tls test-etcd --cert /tmp/test/pki/etcd/ca.crt --key /tmp/test/pki/etcd/ca.key
+kubectl -n workload-api create secret tls "test-apiserver-etcd-client" \
+  --cert "/tmp/test/pki/apiserver-etcd-client.crt" \
+  --key "/tmp/test/pki/apiserver-etcd-client.key"
+kubectl -n workload-api create secret tls apiserver \
+  --cert "/tmp/test/pki/apiserver.crt" \
+  --key "/tmp/test/pki/apiserver.key"
+kubectl -n workload-api create secret generic test-sa \
+  --from-file=tls.crt="/tmp/test/pki/sa.pub" \
+  --from-file=tls.key="/tmp/test/pki/sa.key"
+```
+
+With all that out of the way, we can finally deploy the API server!
+For this we will use a normal Deployment.
+
+```bash
+# Deploy API server
+curl -L https://github.com/Nordix/metal3-clusterapi-docs/raw/main/metal3-scaling-experiments/manifests/v2/kube-apiserver-deployment.yaml |
+  sed "s/CLUSTER/test/g" | kubectl -n workload-api apply -f -
+kubectl -n workload-api wait --for=condition=Available deploy/test-kube-apiserver
+```
+
+Time to check if it worked!
+We can use port-forwarding to access the API, but of course we will need some authentication method for it to be useful.
+With kubeadm we can generate a kubeconfig based on the certificates we already have.
+
+```bash
+kubeadm kubeconfig user --client-name kubernetes-admin --org system:masters \
+  --config tmp/kubeadm-config.yaml > kubeconfig.yaml
```

Comment:
check comment on the top the `kubeadm-config.yaml` is expected to be in the tmp file can we mention it when creating the file

Reply:
Yes this tmp directory seems to have come out of nowhere

Reply:
Well spotted. This `tmp/` was not supposed to be there.